SU2_CFD/src/solvers/CEulerSolver.cpp

```diff
@@ -3122,28 +3125,60 @@ void CEulerSolver::Upwind_Residual(CGeometry *geometry, CSolver **solver_contain
 
       for (iVar = 0; iVar < nPrimVarGrad; iVar++) {
 
-        su2double Project_Grad_i = 0.0;
-        su2double Project_Grad_j = 0.0;
+        const su2double V_ij = V_j[iVar] - V_i[iVar];
+
+        su2double Project_Grad_i = 0.5*Kappa*V_ij;
+        su2double Project_Grad_j = 0.5*Kappa*V_ij;
 
         for (iDim = 0; iDim < nDim; iDim++) {
-          Project_Grad_i += Vector_ij[iDim]*Gradient_i[iVar][iDim];
-          Project_Grad_j -= Vector_ij[iDim]*Gradient_j[iVar][iDim];
+          Project_Grad_i += (1.0-Kappa)*Vector_ij[iDim]*Gradient_i[iVar][iDim];
+          Project_Grad_j += (1.0-Kappa)*Vector_ij[iDim]*Gradient_j[iVar][iDim];
         }
 
         if (limiter) {
           if (van_albada) {
-            su2double V_ij = V_j[iVar] - V_i[iVar];
-            Limiter_i[iVar] = V_ij*( 2.0*Project_Grad_i + V_ij) / (4*pow(Project_Grad_i, 2) + pow(V_ij, 2) + EPS);
-            Limiter_j[iVar] = V_ij*(-2.0*Project_Grad_j + V_ij) / (4*pow(Project_Grad_j, 2) + pow(V_ij, 2) + EPS);
+            Project_Grad_i *= V_ij*(2.0*Project_Grad_i + V_ij) / (4*pow(Project_Grad_i, 2) + pow(V_ij, 2) + EPS);
+            Project_Grad_j *= V_ij*(2.0*Project_Grad_j + V_ij) / (4*pow(Project_Grad_j, 2) + pow(V_ij, 2) + EPS);
           }
-          Primitive_i[iVar] = V_i[iVar] + Limiter_i[iVar]*Project_Grad_i;
-          Primitive_j[iVar] = V_j[iVar] + Limiter_j[iVar]*Project_Grad_j;
-        }
-        else {
-          Primitive_i[iVar] = V_i[iVar] + Project_Grad_i;
-          Primitive_j[iVar] = V_j[iVar] + Project_Grad_j;
-        }
+          else if (piperno) {
+            const su2double R_i = 0.5*V_ij/(Project_Grad_i+EPS);
+            const su2double R_j = 0.5*V_ij/(Project_Grad_j+EPS);
+
+            const su2double InvR_i = (Project_Grad_i)/(0.5*V_ij+EPS);
+            const su2double InvR_j = (Project_Grad_j)/(0.5*V_ij+EPS);
+            
+            Project_Grad_i *= 1.0/6.0*(1.0+2.0*R_i);
+            Project_Grad_j *= 1.0/6.0*(1.0+2.0*R_j);
+
+            if (R_i < 0.0) {
+              Project_Grad_i = 0.0;
+            }
+            else if (R_i < 1.0) {
+              Project_Grad_i *= (3.0*pow(InvR_i, 2.0) - 6.0*InvR_i + 19.0)
+                               / (pow(InvR_i, 3.0) - 3.0*InvR_i + 18.0);
+            }
+            else {
+              Project_Grad_i *= 1.0 + (1.5*InvR_i + 1.0)*pow(InvR_i - 1.0, 3.0);
+            }
 
+            if (R_j < 0.0) {
+              Project_Grad_j = 0.0;
+            }
+            else if (R_j < 1.0) {
+              Project_Grad_j *= (3.0*pow(InvR_j, 2.0) - 6.0*InvR_j + 19.0)
+                               / (pow(InvR_j, 3.0) - 3.0*InvR_j + 18.0);
+            }
+            else {
+              Project_Grad_j *= 1.0 + (1.5*InvR_j + 1.0)*pow(InvR_j - 1.0, 3.0);
+            }
```

Comment:
Please move this piece-wise function to a limiter details class (CLimiterDetails.hpp) as right now it is duplicated here and again for the incompressible solver.
Or make it one of the helper functions in that same file.

Side note, any chance this can be written without branching? Otherwise I'll have a bad time implementing it in #1022.

Reply:
I moved the edge-based limiters to CLimiterDetails.hpp.

> Side note, any chance this can be written without branching?
I'm not sure how to do that since the function is piece-wise. Any suggestions?